hibernate-core/src/main/java/org/hibernate/query/DynamicSelection.java

```diff
@@ -0,0 +1,43 @@
+/*
+ * SPDX-License-Identifier: Apache-2.0
+ * Copyright Red Hat Inc. and Hibernate Authors
+ */
+package org.hibernate.query;
+
+import org.hibernate.Incubating;
+import org.hibernate.query.restriction.Restriction;
+
+/**
+ * A dynamic builder for {@linkplain SelectionQuery} instances.
+ * Once all {@linkplain #addOrder sorting} and {@linkplain #addRestriction restrictions}
+ * are defined, call {@linkplain #createQuery()} to obtain the executable form.
+ *
+ * @see QueryProducer#createDynamicSelection(String, Class)
+ *
+ * @author Steve Ebersole
+ */
+@Incubating
+public interface DynamicSelection<T> {
+	/**
+	 * Adds a sort specification to the in-flight selection definition.
+	 *
+	 * @param order The sorting fragment to be added.
+	 *
+	 * @return {@code this} for method chaining.
+	 */
+	DynamicSelection<T> addOrder(Order<T> order);
```

Comment:
I'm not sure, but we might also want:

```java
DynamicSelection<T> setOrder(Order<T> order);
```

to completely replace the existing order by clause.

Reply:
That's fine.